File: server/src/dataProcessing.ts
import { toJSON } from "danfojs-node";
import { DataFrame, merge } from "danfojs-node/dist/danfojs-base";
import Quaternion from 'quaternion';

interface SensorData {
  time: number;
  x?: number;
  y?: number;
  z?: number;
  qx?: number;
  qy?: number;
  qz?: number;
  qw?: number;
  latitude?: number;
  longitude?: number;
  altitude?: number;
  speed?: number;
}

interface ShockData {
  time: number;
  zAccel: number;
  latitude: number;
  longitude: number;
  altitude: number;
}

// Parse JSON data into an array of SensorData objects
export function parseJSONData(data: any): SensorData[] {
  return data.map((entry: any) => ({
    time: Number(entry.time),
    x: entry.x ? parseFloat(entry.x) : undefined,
    y: entry.y ? parseFloat(entry.y) : undefined,
    z: entry.z ? parseFloat(entry.z) : undefined,
    qx: entry.qx ? parseFloat(entry.qx) : undefined,
    qy: entry.qy ? parseFloat(entry.qy) : undefined,
    qz: entry.qz ? parseFloat(entry.qz) : undefined,
    qw: entry.qw ? parseFloat(entry.qw) : undefined,
    latitude: entry.latitude ? parseFloat(entry.latitude) : undefined,
    longitude: entry.longitude ? parseFloat(entry.longitude) : undefined,
    altitude: entry.altitude ? parseFloat(entry.altitude) : undefined,
    speed: entry.speed ? parseFloat(entry.speed) : undefined,
  }));
}

export function processSensorData(data: SensorData[]): DataFrame {
  const accelData = data
    .filter(d => d.z !== undefined)
    .map(d => ({ time: d.time, z: d.z }));
  const orientData = data
    .filter(d => d.qw !== undefined)
    .map(d => ({ time: d.time, qx: d.qx, qy: d.qy, qz: d.qz, qw: d.qw }));
  const locationData = data
    .filter(d => d.latitude !== undefined)
    .map(d => ({ time: d.time, latitude: d.latitude, longitude: d.longitude, altitude: d.altitude }));

  const accelDf = new DataFrame(accelData);
  const orientDf = new DataFrame(orientData);
  const locationDf = new DataFrame(locationData);

  // Combine DataFrames by 'time' using a left join method
  const mergedData = merge({ left: accelDf, right: orientDf, on: ['time'], how: 'left' });
  const finalMergedData = merge({ left: mergedData, right: locationDf, on: ['time'], how: 'left' });

  return finalMergedData;
}
// Detect shocks in merged data based on the Z acceleration and threshold
export function detectShocks(mergedData: DataFrame, shockThreshold: number = 3): ShockData[] {
  const shocks: ShockData[] = [];

  const jsonData = toJSON(mergedData) as any[];

  jsonData.forEach((row: { [key: string]: any }) => {
    const accelZ = row['z'];
    const qw = row['qw'];
    const qx = row['qx'];
    const qy = row['qy'];
    const qz = row['qz'];

    // Create quaternion and convert to rotation matrix
    const quaternion = new Quaternion(qw, qx, qy, qz);
    const rotationMatrix = quaternion.toMatrix(true) as number[][];

    // Apply rotation to acceleration vector along Z-axis
    const globalZAccel = rotationMatrix[2][0] * 0 + rotationMatrix[2][1] * 0 + rotationMatrix[2][2] * accelZ;

    // Detect shock based on threshold and store shock data
    if (Math.abs(globalZAccel) > shockThreshold) {
      shocks.push({
        time: row['time'],
        zAccel: globalZAccel,
        latitude: row['latitude'],
        longitude: row['longitude'],
        altitude: row['altitude'],
      });
    }
  });

  return shocks;
}

File: server/data_processing.py
from typing import List, Optional
from pydantic import BaseModel
import numpy as np
from scipy.spatial.transform import Rotation as R

class ShockData(BaseModel):
    time: float
    zAccel: float
    latitude: float
    longitude: float
    altitude: float


class SensorData(BaseModel):
    time: float
    x: Optional[float]
    y: Optional[float]
    z: Optional[float]
    qx: Optional[float]
    qy: Optional[float]
    qz: Optional[float]
    qw: Optional[float]
    latitude: Optional[float]
    longitude: Optional[float]
    altitude: Optional[float]
    speed: Optional[float]


def parse_json_data(data: List[dict]) -> List[SensorData]:
    return [SensorData(**entry) for entry in data]


def process_sensor_data(data: List[SensorData]):
    accel_data = [{"time": d.time, "z": d.z} for d in data if d.z is not None]
    orient_data = [{"time": d.time, "qx": d.qx, "qy": d.qy, "qz": d.qz, "qw": d.qw} for d in data if d.qw is not None]
    location_data = [{"time": d.time, "latitude": d.latitude, "longitude": d.longitude, "altitude": d.altitude} for d in data if d.latitude is not None]

    accel_df = {entry["time"]: entry for entry in accel_data}
    orient_df = {entry["time"]: entry for entry in orient_data}
    location_df = {entry["time"]: entry for entry in location_data}

    merged_data = []
    for time in accel_df.keys():
        merged_entry = {**accel_df.get(time, {}), **orient_df.get(time, {}), **location_df.get(time, {})}
        merged_data.append(merged_entry)
    return merged_data


def detect_shocks(merged_data: List[dict], shock_threshold: float = 3.0) -> List[ShockData]:
    shocks = []
    for row in merged_data:
        accel_z = row.get("z", 0)
        qw, qx, qy, qz = row.get("qw", 1), row.get("qx", 0), row.get("qy", 0), row.get("qz", 0)

        rotation = R.from_quat([qx, qy, qz, qw])
        global_z_accel = rotation.apply([0, 0, accel_z])[2]

        if abs(global_z_accel) > shock_threshold:
            shocks.append(ShockData(
                time=row["time"],
                zAccel=global_z_accel,
                latitude=row.get("latitude", 0),
                longitude=row.get("longitude", 0),
                altitude=row.get("altitude", 0)
            ))
    return shocks
